File: api/v1/schemas/alerts.py
# -*- coding: utf-8 -*-
"""Alert API schemas."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

from api.v1.schemas.history import AnalysisContextPackOverview
from api.v1.schemas.market_phase import MarketPhaseSummary


TargetScopeValue = Literal["single_symbol", "watchlist", "portfolio_holdings", "portfolio_account", "market"]
SeverityValue = Literal["info", "warning", "critical"]
DryRunStatusValue = Literal["triggered", "not_triggered", "evaluation_error"]
TargetRecordStatusValue = Literal["triggered", "skipped", "degraded", "failed"]


class AlertRuleCreateRequest(BaseModel):
    name: Optional[str] = Field(None, max_length=64)
    target_scope: TargetScopeValue = "single_symbol"
    target: str = Field(..., min_length=1, max_length=64)
    alert_type: str = Field(..., min_length=1, max_length=32)
    parameters: Dict[str, Any] = Field(default_factory=dict)
    severity: SeverityValue = "warning"
    enabled: bool = True
    cooldown_policy: Optional[Dict[str, Any]] = None
    notification_policy: Optional[Dict[str, Any]] = None


class AlertRuleUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, max_length=64)
    target_scope: Optional[TargetScopeValue] = None
    target: Optional[str] = Field(None, min_length=1, max_length=64)
    alert_type: Optional[str] = Field(None, min_length=1, max_length=32)
    parameters: Optional[Dict[str, Any]] = None
    severity: Optional[SeverityValue] = None
    enabled: Optional[bool] = None
    cooldown_policy: Optional[Dict[str, Any]] = None
    notification_policy: Optional[Dict[str, Any]] = None


class AlertRuleItem(BaseModel):
    id: int
    name: str
    target_scope: str
    target: str
    alert_type: str
    parameters: Dict[str, Any] = Field(default_factory=dict)
    severity: str
    enabled: bool
    source: str
    cooldown_policy: Optional[Dict[str, Any]] = None
    notification_policy: Optional[Dict[str, Any]] = None
    last_triggered_at: Optional[str] = None
    cooldown_until: Optional[str] = None
    cooldown_active: Optional[bool] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class AlertRuleListResponse(BaseModel):
    items: List[AlertRuleItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class AlertDeleteResponse(BaseModel):
    deleted: int


class AlertRuleTargetResult(BaseModel):
    target: str
    display_target: Optional[str] = None
    status: DryRunStatusValue
    record_status: Optional[TargetRecordStatusValue] = None
    triggered: bool
    observed_value: Optional[Any] = None
    threshold: Optional[Any] = None
    message: str


class AlertRuleTestResponse(BaseModel):
    rule_id: int
    target_scope: Optional[str] = None
    status: DryRunStatusValue
    triggered: bool
    observed_value: Optional[Any] = None
    message: str
    evaluated_count: int = 0
    triggered_count: int = 0
    degraded_count: int = 0
    skipped_count: int = 0
    target_results: List[AlertRuleTargetResult] = Field(default_factory=list)


class AlertTriggerItem(BaseModel):
    id: int
    rule_id: Optional[int] = None
    target: str
    observed_value: Optional[float] = None
    threshold: Optional[float] = None
    reason: Optional[str] = None
    data_source: Optional[str] = None
    data_timestamp: Optional[str] = None
    triggered_at: Optional[str] = None
    status: str
    diagnostics: Optional[str] = None
    market_phase_summary: Optional[MarketPhaseSummary] = None
    analysis_context_pack_overview: Optional[AnalysisContextPackOverview] = None
    analysis_visibility_source: Optional[str] = Field(
        None,
        description=(
            "公开摘要来源：alert_trigger_market_context / analysis_history_snapshot / "
            "evaluator_snapshot / legacy_text / null"
        ),
    )
    decision_signal_summary: Optional[Dict[str, Any]] = None


class AlertTriggerListResponse(BaseModel):
    items: List[AlertTriggerItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class AlertNotificationItem(BaseModel):
    id: int
    trigger_id: Optional[int] = None
    channel: str
    attempt: int
    success: bool
    error_code: Optional[str] = None
    retryable: bool
    latency_ms: Optional[int] = None
    diagnostics: Optional[str] = None
    created_at: Optional[str] = None


class AlertNotificationListResponse(BaseModel):
    items: List[AlertNotificationItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int

File: apps/dsa-web/src/types/alerts.ts
import type { AnalysisContextPackOverview, MarketPhaseSummary } from './analysis';
import type { DecisionSignalItem } from './decisionSignals';

export type AlertType =
  | 'price_cross'
  | 'price_change_percent'
  | 'volume_spike'
  | 'ma_price_cross'
  | 'rsi_threshold'
  | 'macd_cross'
  | 'kdj_cross'
  | 'cci_threshold'
  | 'portfolio_stop_loss'
  | 'portfolio_concentration'
  | 'portfolio_drawdown'
  | 'portfolio_price_stale'
  | 'market_light_status'
  | 'market_light_score_drop';
export type AlertSeverity = 'info' | 'warning' | 'critical';
export type AlertTargetScope = 'single_symbol' | 'watchlist' | 'portfolio_holdings' | 'portfolio_account' | 'market';
export type AlertDirection = 'above' | 'below' | 'up' | 'down' | 'bullish_cross' | 'bearish_cross';
export type PortfolioStopLossMode = 'near' | 'breach';
export type MarketRegion = 'cn' | 'hk' | 'us';
export type MarketLightStatus = 'yellow' | 'red';
export type AlertDryRunStatus = 'triggered' | 'not_triggered' | 'evaluation_error';
export type AlertTriggerStatus = 'triggered' | 'skipped' | 'degraded' | 'failed';

export interface AlertRuleParameters {
  direction?: AlertDirection;
  price?: number;
  changePct?: number;
  multiplier?: number;
  window?: number;
  period?: number;
  threshold?: number;
  fastPeriod?: number;
  slowPeriod?: number;
  signalPeriod?: number;
  kPeriod?: number;
  dPeriod?: number;
  mode?: PortfolioStopLossMode;
  statuses?: MarketLightStatus[];
  minDrop?: number;
}

export interface AlertRuleItem {
  id: number;
  name: string;
  targetScope: AlertTargetScope;
  target: string;
  alertType: AlertType;
  parameters: AlertRuleParameters;
  severity: AlertSeverity;
  enabled: boolean;
  source: string;
  cooldownPolicy?: Record<string, unknown> | null;
  notificationPolicy?: Record<string, unknown> | null;
  lastTriggeredAt?: string | null;
  cooldownUntil?: string | null;
  cooldownActive?: boolean | null;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface AlertRuleListResponse {
  items: AlertRuleItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface AlertRuleCreateRequest {
  name?: string;
  targetScope?: AlertTargetScope;
  target: string;
  alertType: AlertType;
  parameters: AlertRuleParameters;
  severity: AlertSeverity;
  enabled?: boolean;
}

export interface AlertDeleteResponse {
  deleted: number;
}

export interface AlertRuleTestResponse {
  ruleId: number;
  targetScope?: AlertTargetScope | string | null;
  status: AlertDryRunStatus;
  triggered: boolean;
  observedValue?: unknown;
  message: string;
  evaluatedCount?: number;
  triggeredCount?: number;
  degradedCount?: number;
  skippedCount?: number;
  targetResults?: AlertRuleTargetResult[];
}

export interface AlertRuleTargetResult {
  target: string;
  displayTarget?: string | null;
  status: AlertDryRunStatus;
  recordStatus?: AlertTriggerStatus | null;
  triggered: boolean;
  observedValue?: unknown;
  threshold?: unknown;
  message: string;
}

export interface AlertTriggerItem {
  id: number;
  ruleId?: number | null;
  target: string;
  observedValue?: number | null;
  threshold?: number | null;
  reason?: string | null;
  dataSource?: string | null;
  dataTimestamp?: string | null;
  triggeredAt?: string | null;
  status: AlertTriggerStatus | string;
  diagnostics?: string | null;
  marketPhaseSummary?: MarketPhaseSummary | null;
  analysisContextPackOverview?: AnalysisContextPackOverview | null;
  analysisVisibilitySource?: string | null;
  decisionSignalSummary?: Partial<DecisionSignalItem> | null;
}

export interface AlertTriggerListResponse {
  items: AlertTriggerItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface AlertNotificationItem {
  id: number;
  triggerId?: number | null;
  channel: string;
  attempt: number;
  success: boolean;
  errorCode?: string | null;
  retryable: boolean;
  latencyMs?: number | null;
  diagnostics?: string | null;
  createdAt?: string | null;
}

export interface AlertNotificationListResponse {
  items: AlertNotificationItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface AlertRuleListQuery {
  enabled?: boolean;
  alertType?: AlertType;
  targetScope?: AlertTargetScope;
  target?: string;
  source?: string;
  page?: number;
  pageSize?: number;
}

export interface AlertTriggerListQuery {
  ruleId?: number;
  target?: string;
  status?: string;
  page?: number;
  pageSize?: number;
}

export interface AlertNotificationListQuery {
  triggerId?: number;
  channel?: string;
  success?: boolean;
  page?: number;
  pageSize?: number;
}
